### PR TITLE
Form Features

### PR DESCRIPTION
**Pull request by george-shammar**
This pull request seeks to add the following features to the development branch:
1. HTML form to create new users. This has been commented out as directed in the project instruction.
2. create, new, edit action were created in UserController with the corresponding routes.
3. Nested users hash with params.
4. Top level users params (commented out as directed in the project instructions)
5. Rails' form_with to create new user as well as to edit user.
6. Error messages to provide feedback to users when validation is failed.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  def new
+    @user = User.new
+  end
+
+  def edit
+    @user = User.find_by(id: params[:id])
+  end
+
+  def create
+    # @user = User.new(username: params[:username], email: params[:email], password: params[:password])
```
Comment (lekegitrepo) [inline app/controllers/users_controller.rb]: - Remove commented code

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @george-shammar and @pipe2442 :clap: :clap:
#### Good job so far :+1: It's easy to understand how you implement this project which is a good thing as a developer.
#### However, some requirements are missing/not implemented correctly. Check my comments below :point_down:

#### 1) Readme:
 - In the readme instructions:
     - Add instruction on how to install the project dependencies
     - Specifically, instruct the user to add this command as part of the dependencies installation `yarn install`

#### 2) App Behaviour:
 - Users should be able to edit their details in the app by using this URL `http://localhost:3000/users/1/edit` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/edit_user.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms)). This URL will pick the specified user(in this case user with id '1') and prefill the user's detail in the edit form and the user can make changes and save/submit

#### 3) User Controller:
 -  `update action`
    * [x] It should find user based on id from params, i.e. `@user = User.find(params[:id])` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_params.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))
    * [ ] It should `redirect_to edit_user_path` if user was saved
    * [x] It should `render :edit` if user was not saved ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_update.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

#### 4) Routes:
 - In `config/routes.rb` you should specify `resources :users, only: [:new, :create, :edit, :update]` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route1.png) [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms)).

#### Forms:
 - Commented `Html form`:
     - uses `users_path` helper for action
 - Commented `form_tag`:
     - uses `label_tag`, `text_field_tag`, `email_field_tag`, `password_field_tag`

#### Optional:
 - You can create partial for user form and reused in edit and new view.
 - You can validate the email format by using either a regular expression or a built in URI method.

#### _Reviewed by Temitayo Adeleke_

File: README.md [hunk 2]
```diff
@@ -1 +1,78 @@
...
+- [Ruby On Rails](https://rubyonrails.org/).
+- [Git](https://git-scm.com/).
+
+## Getting Started
+
+- **[Install Ruby on your machine.](https://www.theodinproject.com/courses/ruby-programming/lessons/installing-ruby-ruby-programming)** 
+- **[Install Rails on your machine.](https://www.theodinproject.com/courses/ruby-on-rails/lessons/your-first-rails-application-ruby-on-rails)**
+- **[Install Git on your machine.](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)**
+- Install Node.js and Yarn. According to the official yarn website, it is recommended to install Yarn through the npm package manager, which comes bundled with Node.js when you install it on your system.
+Once you have npm installed you can run the following both to install and upgrade Yarn:
```
Comment (whiz25) [inline README.md:31]: - What command should I run to both install and upgrade `yarn`? Be explicit in your instructions.

File: README.md [hunk 3]
```diff
@@ -1 +1,78 @@
...
+- [rspec](https://rspec.info/).
+
+
+### Usage
+
+- Create new folder.
+- Open your terminal.
+- Run ``` git clone https://github.com/pipe2442/re-former.git ```.
+- Enter root folder with command ``` cd re-former ```.
+- In the terminal from the root folder, type ``` rails s ``` and run the browser to display the project.
```
Comment (whiz25) [inline README.md]: - Before starting the server, doesn't a user need to install dependencies and migrate the database? Add instructions for checking out the `feature/re-former` branch and installing `gems`

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 
Good job team @george-shammar  and @pipe2442 . You did well addressing some of the concerns raised by the previous user. Comments were removed from the codebase, the `users_controller` updated accordingly. 
However, the `README` still lacks clear instructions for starting this app locally. Check my comments for details.


### Martin Afani

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello George Gbenle & Felipe Angulo Torres** 
Good job with the improvements for this project :muscle:  :muscle:
Your user's actions are working perfectly :clap: :clap: :clap:  
There are just a couple of details to improve before you can go ahead:

- [x] You might edit rubocop to exclude folders like node_modules or config
- [ ] Add screenshot to the README file
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by MarilenaRoque — APPROVED** [review]
## Hello Team! :rocket:

Congratulations on achieving all the requirements for this project.

# Your Project is Approved! 🥇 

## To Highlight 🌟 
- Pull Request message really descriptive. Good Job!

Keep going with the hard work! 💪🏻 
Happy Coding! 💻 

**_Marilena Roque
Full-Stack Developer
Code Reviewer at Microverse Inc_.**

**Pull request merged**